### PR TITLE
Change Saving games documentation to use ConfigFile, JSON documentation moved to separate file

### PR DESCRIPTION
**Pull request by ShatReal**
Using JSON in the only saving tutorial is misleading and leads users to try and save everything in JSON when it's not necessary. I've changed the code to use ConfigFile instead.

However, I don't know C# so someone else will have to write the C# code for the Saving games file (JSON already written).

### CODE REVIEW

**Comment by HaSa1002** [comment]
You can use my [code translator](https://github.com/HaSa1002/codetranslator), to get an inital translation. I can review the code once you pushed it.

**Comment by ShatReal** [comment]
Okay, it's finally passed all check successfully! 😄

File: tutorials/io/saving_games.rst [hunk 1]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
...
 
     // Note: This can be called from anywhere inside the tree. This function is
     // path independent.
     public void LoadGame()
     {
-        var saveGame = new File();
-        if (!saveGame.FileExists("user://savegame.save"))
-            return; // Error!  We don't have a save to load.
+        var saveGame = new ConfigFile();
+        var err =SaveGame.load("user://savegame.save") = saveGame.GD.Load("user://savegame.save");
```
Comment (clemens-tolboom) [inline tutorials/io/saving_games.rst:223]: This looks weird loading twice. Copy/paste err?

Comment (ShatReal) [reply tutorials/io/saving_games.rst:223]: Yeah. ConfigFiles don't have to check if the file exists. I don't know C# but I'll try to fix it

File: tutorials/io/saving_games.rst [hunk 2]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
...
     // Note: This can be called from anywhere inside the tree. This function is
     // path independent.
     public void LoadGame()
     {
-        var saveGame = new File();
-        if (!saveGame.FileExists("user://savegame.save"))
-            return; // Error!  We don't have a save to load.
+        var saveGame = new ConfigFile();
+        var err = saveGame.Load("user://savegame.save");
+        if (err != OK)
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:224]: Hey 🙂 You're probably looking for `Error.Ok` instead of `OK`.

File: tutorials/io/saving_games.rst [hunk 3]
```diff
@@ -25,25 +20,22 @@ sessions and what information we want to keep from those objects. For
...
 
-    var save_nodes = get_tree().get_nodes_in_group("Persist")
+    var save_nodes = get_tree().get_nodes_in_group("persist")
     for i in save_nodes:
         # Now, we can call our save function on each node.
 
  .. code-tab:: csharp
 
-    var saveNodes = GetTree().GetNodesInGroup("Persist");
+    var saveNodes = GetTree().GetNodesInGroup("persist");
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:38]:
```suggestion
    Godot.Collections.Array saveNodes = GetTree().GetNodesInGroup("persist");
```

File: tutorials/io/saving_games.rst [hunk 4]
```diff
@@ -55,36 +47,23 @@ Serializing
...
-            "experience" : experience,
-            "tnl" : tnl,
-            "level" : level,
-            "attack_growth" : attack_growth,
-            "defense_growth" : defense_growth,
-            "health_growth" : health_growth,
-            "is_alive" : is_alive,
-            "last_attack" : last_attack
+            "name": name,
+            "filename": get_filename(),
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:60]:
```suggestion
            "filename": Filename,
```

File: tutorials/io/saving_games.rst [hunk 5]
```diff
@@ -168,10 +133,9 @@ way to pull the data out of the file as well.
     // dict of relevant variables.
     public void SaveGame()
     {
-        var saveGame = new File();
-        saveGame.Open("user://savegame.save", (int)File.ModeFlags.Write);
+        var saveGame = new ConfigFile();
 
-        var saveNodes = GetTree().GetNodesInGroup("Persist");
+        var saveNodes = GetTree().GetNodesInGroup("persist");
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:138]:
```suggestion
        Godot.Collections.Array saveNodes = GetTree().GetNodesInGroup("persist");
```

File: tutorials/io/saving_games.rst [hunk 6]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
             // Call the node's save function.
             var nodeData = saveNode.Call("Save");
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:156]:
```suggestion
            var nodeData = (Godot.Collections.Dictionary)saveNode.Call("Save");
```

File: tutorials/io/saving_games.rst [hunk 7]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
...
 
-            // Store the save dictionary as a new line in the save file.
-            saveGame.StoreLine(JSON.Print(nodeData));
+            // Iterates through save dict and stores each key-value pair under the node's name's section
+            foreach (var Key in NodeData)
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:159]:
```suggestion
            foreach (string key in nodeData.Keys)
```

File: tutorials/io/saving_games.rst [hunk 8]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
...
+            {
+                saveGame.SetValue(NodeData["name"], Key, NodeData[Key]);
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:161]:
```suggestion
                saveGame.SetValue(nodeData["Name"], key, nodeData[key]);
```

File: tutorials/io/saving_games.rst [hunk 9]
```diff
@@ -191,67 +155,76 @@ way to pull the data out of the file as well.
...
 
     // Note: This can be called from anywhere inside the tree. This function is
     // path independent.
     public void LoadGame()
     {
-        var saveGame = new File();
-        if (!saveGame.FileExists("user://savegame.save"))
-            return; // Error!  We don't have a save to load.
+        var saveGame = new ConfigFile();
+        var err = saveGame.Load("user://savegame.save");
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:223]:
```suggestion
        Error err = saveGame.Load("user://savegame.save");
```

File: tutorials/io/saving_games.rst [hunk 10]
```diff
@@ -261,32 +234,32 @@ load function:
         foreach (Node saveNode in saveNodes)
             saveNode.QueueFree();
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:235]:
```suggestion
        {
            saveNode.QueueFree();
        }
```

Comment (paulloz) [reply tutorials/io/saving_games.rst:235]: I can't annotate l.233, but it should be
```csharp
Godot.Collections.Array saveNodes = GetTree().GetNodesInGroup("persist");
```

File: tutorials/io/saving_games.rst [hunk 11]
```diff
@@ -261,32 +234,32 @@ load function:
...
 
-        // Load the file line by line and process that dictionary to restore the object
-        // it represents.
-        saveGame.Open("user://savegame.save", (int)File.ModeFlags.Read);
-
-        while (saveGame.GetPosition() < saveGame.GetLen())
+        // Process each section in the file to restore the object it represents.
+        foreach (var Section in SaveGame.GetSections())
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:238]:
```suggestion
        foreach (string section in saveGame.GetSections())
```

File: tutorials/io/saving_games.rst [hunk 12]
```diff
@@ -261,32 +234,32 @@ load function:
...
         {
-            // Get the saved dictionary from the next line in the save file
-            var nodeData = new Godot.Collections.Dictionary<string, object>((Godot.Collections.Dictionary)JSON.Parse(saveGame.GetLine()).Result);
+            var nodeData = {};
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:240]:
```suggestion
            var nodeData = new Godot.Collections.Dictionary<string, object>();
```

File: tutorials/io/saving_games.rst [hunk 13]
```diff
@@ -261,32 +234,32 @@ load function:
...
+            // Adds each key in the section to the dictionary.
+            foreach (var Key in SaveGame.GetSectionKeys(Section))
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:242]:
```suggestion
            foreach (string key in saveGame.GetSectionKeys(section))
```

File: tutorials/io/saving_games.rst [hunk 14]
```diff
@@ -261,32 +234,32 @@ load function:
...
+            {
+                nodeData[Key] = SaveGame.GetValue(Section, Key);
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:244]:
```suggestion
                nodeData[key] = saveGame.GetValue(section, key);
```

File: tutorials/io/saving_games.rst [hunk 15]
```diff
@@ -261,32 +234,32 @@ load function:
...
+            }
 
+        {
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:247]:
```suggestion
```

File: tutorials/io/saving_games.rst [hunk 16]
```diff
@@ -261,32 +234,32 @@ load function:
...
             // Firstly, we need to create the object and add it to the tree and set its position.
             var newObjectScene = (PackedScene)ResourceLoader.Load(nodeData["Filename"].ToString());
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:249]:
```suggestion
            var newObjectScene = ResourceLoader.Load<PackedScene>(nodeData["Filename"].ToString());
```

File: tutorials/io/saving_games.rst [hunk 17]
```diff
@@ -261,32 +234,32 @@ load function:
...
             var newObject = (Node)newObjectScene.Instance();
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:250]:
```suggestion
            Node newObject = newObjectScene.Instance();
```

File: tutorials/io/saving_games.rst [hunk 18]
```diff
@@ -261,32 +234,32 @@ load function:
...
             GetNode(nodeData["Parent"].ToString()).AddChild(newObject);
-            newObject.Set("Position", new Vector2((float)nodeData["PosX"], (float)nodeData["PosY"]));
+            newObject.Set("Position", nodeData["Pos"]));
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:252]:
```suggestion
            newObject.Set("position", nodeData["Pos"]));
```

File: tutorials/io/saving_games.rst [hunk 19]
```diff
@@ -261,32 +234,32 @@ load function:
...
 
             // Now we set the remaining variables.
             foreach (KeyValuePair<object, object> entry in nodeData)
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:255]:
```suggestion
            foreach (KeyValuePair<string, object> entry in nodeData)
```

File: tutorials/io/saving_games.rst [hunk 20]
```diff
@@ -261,32 +234,32 @@ load function:
...
             {
                 string key = entry.Key.ToString();
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:257]:
```suggestion
```

File: tutorials/io/saving_games.rst [hunk 21]
```diff
@@ -261,32 +234,32 @@ load function:
...
-                if (key == "Filename" || key == "Parent" || key == "PosX" || key == "PosY")
+                if (key == "Filename" || key == "Parent" || key == "Pos")
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:258]:
```suggestion
                if (entry.Key == "Filename" || entry.Key == "Parent" || entry.Key == "Pos")
```

File: tutorials/io/saving_games.rst [hunk 22]
```diff
@@ -261,32 +234,32 @@ load function:
...
                     continue;
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:259]:
```suggestion
                {
                    continue;
                }
```

File: tutorials/io/saving_games.rst [hunk 23]
```diff
@@ -261,32 +234,32 @@ load function:
...
                 newObject.Set(key, entry.Value);
```
Comment (paulloz) [inline tutorials/io/saving_games.rst:260]:
```suggestion
                newObject.Set(entry.Key, entry.Value);
```

**Review by paulloz — COMMENTED** [review]
Hi, feel free to @ me if anything is unclear. Submitting changes this way isn't the most practical 😅
A lot of those changes should also be applied to the *Saving with JSON* page. I can take time to highlight them a bit later if need be.

**Comment by ShatReal** [comment]
So I tried squashing my commits and this happened.

**Comment by ShatReal** [comment]
oh no help

**Comment by pycbouh** [comment]
You seem to have squashed your commit and all of the history of the `master` branch into one commit... The PR cannot be opened, because it no longer shares the history with our code base.

**Comment by ShatReal** [comment]
Unfortunately the answer I followed had me delete my .git file. Well, I can always recreate the changes in a new pull request.

**Comment by paulloz** [comment]
I have a snapshot of your PR branch on my own fork [here](https://github.com/paulloz/godot-docs/tree/pr-4698).

**Comment by ShatReal** [comment]
Alright cool 👍🏼

**Comment by pycbouh** [comment]
@ShatReal Do you still plan to restore this PR?

**Comment by ShatReal** [comment]
Honestly not yet since KidsCanCode opened that issue about improving the save documentation
I want to come to a conclusion before writing it

**Pull request closed without merging**